### DISCUSSION TITLE
Fedora packaging: Creation of official uftrace rpm for inclusion into Fedora

### DISCUSSION THREAD

**Discussion by bernhardkaindl** [namhyung/uftrace#1650] [General]
Hi @namhyung, I was active with Debian packaging of uftrace in the past, and hope for your support for this:

I'ld like to coordinate with you about creation of an official Fedora rpm for inclusion into Fedora itself.

I've been an rpm packager for many years and I found already two attempts to submit an uftrace rpm to Fedora.

I'd like to follow it thru this time and make it a reality.

Since I am not yet in the Fedora Packagers Group, it would help if people would encourage me and support me when I apply to be sponsored into the Fedora Packages Group.

Many thanks for any participation, and lets hope to get this done!

Best regards,
Bernhard

**Comment by namhyung** [comment]
Hello @bernhardkaindl.

Thanks for reaching out.  I'd be more than happy if you'd work on the Fedora packaging.  Please let me know if anything I can help.

**Reply by bernhardkaindl** [reply]
@namhyung The initial builds are out!

To be checked by you anyone having any Red Hat installs:

https://copr.fedorainfracloud.org/coprs/bkaindl/uftrace/builds/

The `uftrace.spec` file is here: https://github.com/bernhardkaindl/uftrace/blob/master/uftrace.spec

- x86_64 builds look good across the board (tests are enabled), 
- All the test cases of at leat one aarch64 of 0.13 are segfaulting! I've made it so that they pass so you peek the aarch64 test logs.

https://download.copr.fedorainfracloud.org/results/bkaindl/uftrace/fedora-37-aarch64/05681992-uftrace/build.log.gz
```py
Please report this bug to https://github.com/namhyung/uftrace/issues.
WARN: [1] (ns::ns1::foo::foo[aaaadff30c64] <= <70aaaadff30a70>[70aaaadff30a70])
WARN: [0] (main[aaaadff30a5c] <= <70ffff86b6b588>[70ffff86b6b588])
WARN: child terminated by signal: 11: Segmentation fault
WARN: Segmentation fault: address not mapped (addr: 0x48ffffb5e6b0e8)
WARN:  if this happens only with uftrace, please consider -e/--estimate-return option.
WARN: Backtrace from uftrace v0.13 ( aarch64 dwarf python3 luajit tui perf sched dynamic )
WARN: =====================================
Please report this bug to https://github.com/namhyung/uftrace/issues.
WARN: [1] (ns::ns1::foo::foo[aaaab8c20c64] <= <1aaaaab8c20a70>[1aaaaab8c20a70])
WARN: [0] (main[aaaab8c20a5c] <= <62ffffb5a5b588>[62ffffb5a5b588])
WARN: child terminated by signal: 11: Segmentation fault
WARN: Segmentation fault: address not mapped (addr: 0x53ffff9086b0e8)
WARN:  if this happens only with uftrace, please consider -e/--estimate-return option.
WARN: Backtrace from uftrace v0.13 ( aarch64 dwarf python3 luajit tui perf sched dynamic )
...
```
Hints appreciated!

**Reply by namhyung** [reply]
Looks like it's crashing on most test cases.  Probably due to different compiler settings.  Can you please show me the disassembly of the test binary like `t-abc` ?

**Reply by bernhardkaindl** [reply]
Hi @namhyung! I haven't yet reproduced these segfaults on aarch64 locally yet:

Assuming or to exclude that they might be caused by Red Hat/Fedora's special compiler flags, I am now trying to build without them.

I already added `%undefine _hardened_build` to get `-fstack-protector=strong` and maybe more out of the way, but this was not enough, the flags at that point were `-fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS  -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mbranch-protection=standard -fasynchronous-unwind-tables -fstack-clash-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer`.

In the newest build tag, I am now removing all `CFLAGS`m `CXXFLAGS`, `LDFLAGS` flags passed from the RPM macros into the build:
Source/spec: https://github.com/bernhardkaindl/uftrace/blob/uftrace-0.13-5/uftrace.spec
Build infos and logs: https://copr.fedorainfracloud.org/coprs/bkaindl/uftrace/build/5693913/

Test passes in RHEL9/CentOS Stream 9 now! Results:
```py
total  4200  Tests executed (success: 80.62%)
  OK:  3326  Test succeeded
  OK:    60  Test succeeded (with some fixup)
  NG:   236  Different test result
  NZ:    23  Non-zero return value
  SG:    30  Abnormal exit by signal
  TM:     0  Test ran too long
  BI:    15  Build failed
  LA:     0  Unsupported Language
  SK:   510  Skipped
 ```
 Looks much better now!

**Reply by namhyung** [reply]
Nice, I suspect `-fstack-clash-protection` might be the culprit.

**Reply by bernhardkaindl** [reply]
@namhyung I was able to disable the hardening `CFLAGS` of Fedora using the RPM macros for disabling them.

For `aarch64`, the remaining C/CXXFLAGS are:
```py
-O2
-fexceptions
-g -grecord-gcc-switches -pipe -Wall -Werror=format-security
-Wp,-D_GLIBCXX_ASSERTIONS
-specs=/usr/lib/rpm/redhat/redhat-annobin-cc1
-mbranch-protection=standard
-fasynchronous-unwind-tables
-fstack-clash-protection 
```
In addition, I added `unset CFLAGS CXXFLAGS LDFLAGS` before running `make test` in case the test cases end up being compiled there. Builds for Fedora, CentOS Stream 8, 9 and derivatives and are all passing now. Hurray!

There is one change I made to `test/runtest.py`: https://github.com/bernhardkaindl/uftrace/commit/2b3d0d9791bf0aebe004113c0d224b72a1938ce3

I added handling the IndexError from self.sort because on the older distributions, a few tests aborted the entire runtest.py because the IndexError wasn't handled. I don't know the code and intent behind it, but this helped the make test to pass on all distributions.

The current consolidated diff of my changes is: https://github.com/namhyung/uftrace/compare/v0.13.1...bernhardkaindl:uftrace:uftrace-0.13.2-7

The builds and their logs are here: https://copr.fedorainfracloud.org/coprs/bkaindl/uftrace/builds/

**Comment by alexmyczko** [comment]
@bernhardkaindl how to reach you for off-topic questions? are you in irc?

**Reply by bernhardkaindl** [reply]
I'm LinkedIn if you have professional or other inquiries.

If you're not on it, or cannot invite me there, you can always create a private repo and invite me to it. Then we can share non-public information using the private repo and connect. What's that off-topic topic?
